core/src/query.rs

```diff
@@ -1,77 +1,551 @@
+                                    ..Default::default()
+                                }
+                            }
+                            // If one of the BSON types is Object, move the
+                            // properties, required, and additional_properties
+                            // information into the schema that will be nested
+                            // in the top-level any_of and set the top-level
+                            // fields to None.
+                            BsonTypeName::Object => {
+                                let nested_properties = properties.clone();
+                                let nested_required = required.clone();
+                                let nested_additional_properties = additional_properties;
+                                properties = None;
+                                required = None;
+                                additional_properties = false;
+                                SimplifiedJsonSchema {
+                                    bson_type: Some(t),
+                                    properties: nested_properties,
+                                    required: nested_required,
+                                    additional_properties: nested_additional_properties,
+                                    ..Default::default()
+                                }
+                            }
+                            // If we encounter any other BSON type, simply create
+                            // a SimplifiedJsonSchema with the relevant bson_type.
+                            t => SimplifiedJsonSchema {
+                                bson_type: Some(t),
+                                ..Default::default()
+                            },
+                        }
+                    })
+                    .collect::<Vec<SimplifiedJsonSchema>>();
+
+                // Extend the top-level any_of with the new, additional
+                // schemas created here from the Multiple BSON types.
+                any_of = match any_of {
+                    None => Some(additional_any_of_schemas),
+                    Some(mut v) => {
+                        v.append(&mut additional_any_of_schemas);
+                        Some(v)
+                    }
+                };
+
+                // Set the top-level bson_type to None when there are multiple
+                // BSON types.
+                None
+            }
+        };
+
+        // Flatten nested any_ofs
+        any_of = any_of.map(|v| {
+            v.into_iter()
+                .flat_map(|s| s.any_of.unwrap_or(vec![]))
+                .collect::<Vec<SimplifiedJsonSchema>>()
+        });
+
+        SimplifiedJsonSchema {
+            bson_type,
+            properties,
+            any_of,
+            required,
+            items,
+            additional_properties,
+        }
+    }
+}
+
+impl TryFrom<(String, SimplifiedJsonSchema)> for BsonTypeInfo {
```

Comment:
I made this a `TryFrom` on `(String, SimplifiedJsonSchema)` purely so we could have the field name in the method so it could be included in the error message. On the one hand, this makes this less general purpose, but on the other ... this doesn't need to be general purpose since it is really specific to converting a (field_name, field_schema) pair into BsonTypeInfo _for that field_.